metricbeat/mb/module/wrapper.go

```diff
@@ -236,6 +236,13 @@ func (msw *metricSetWrapper) fetch(reporter reporter) {
 	case mb.ReportingMetricSetV2:
 		reporter.StartFetchTimer()
 		fetcher.Fetch(reporter.V2())
+	case mb.ReportingMetricSetV2Error:
+		reporter.StartFetchTimer()
+		err := fetcher.Fetch(reporter.V2())
+		if err != nil {
+			reporter.V2().Error(err)
+			logp.Info("Error fetching data for metricset %s.%s: %s", msw.module.Name(), msw.Name(), err)
```

Comment:
Should errors be logged at `Info` level or is there an `Error` level that should be used?

Reply:
That is the part I'm on the fence. My thinking:

* If Metricbeat logs an error, it should mean the operator needs to take actions. If the connection to a service does not work for a few minutes because it was restarted, it's correct that Metricbeat reports error documents, but should it also log an error? No actions from the operator are needed. If someone is running in k8s with autodiscovery I would expect this to happen even more often but no actions are need from the person responsible for the service. Especially if with hints based autodiscovery we allow "someone else" to configure the module.
* The downside is when setting up a module, it shows it very clearly that for example username / password are wrong without having to look at the INFO log or the events itself.

I'm currently leaning in the direction of having `INFO` log as during the initial setup I would expect someone to potentially have info enabled anyway and only switch to `ERROR` in production to have less noise. As @kaiyan-sheng suggests in https://github.com/elastic/beats/pull/10727#issuecomment-463344070 we could also make it a config option and just have `INFO` level as default.

Reply:
So there are two scenarios here:

1. Temporary errors, when the service is being restarted, etc. I agree that logging such errors as `ERROR` in the log creates noise. For this I'd prefer `INFO` or `WARN` at best.

2. Sometimes a service will change its API, perhaps after an upgrade. In this case I think logging at `ERROR` would be nice.

I'm +1 for providing a parameter per @kaiyan-sheng's suggestion that defaults the logging level to `INFO`. That way, if the module creator wants to distinguish between the two scenarios above (or other scenarios I'm not thinking of currently), they can.

Reply:
Now looking again at the code, I don't see a simple way of adding an additional flag but keeping it as simple as it is.

What a metricset can always do, log on the Error level on top of what is logged. Also a metricset does not have to use the interface with the error but can keep using it without it so we have the same behaviour as before. Don't we allow with this already all the flexibility?

The overall goal is making code simpler for 80% of the cases but don't prevent the 20% to do their "own" thing.

Reply:
Sounds good.

Reply:
We can also add in the future an interface that errors can optionally implement to customize how they are reported and logged (or if they are always reported and logged #9091).